src/NServiceBus.Core.Analyzer/AwaitOrCaptureTasksAnalyzer.cs

```diff
@@ -0,0 +1,90 @@
+namespace NServiceBus.Core.Analyzer
+{
+    using System;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Diagnostics;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class AwaitOrCaptureTasksAnalyzer : DiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(diagnostic);
+
+        public override void Initialize(AnalysisContext context) => context.RegisterSyntaxNodeAction(Analyze, SyntaxKind.InvocationExpression);
+
+        void Analyze(SyntaxNodeAnalysisContext context)
+        {
+            if (!(context.Node is InvocationExpressionSyntax call))
+            {
+                return;
+            }
+
+            if (!(call.Parent is ExpressionStatementSyntax))
+            {
+                return;
+            }
+
+            foreach (var token in call.Expression?.DescendantTokens() ?? Enumerable.Empty<SyntaxToken>())
```

Comment:
From some benchmarking it would be slightly faster to use a null check to skip the foreach instead of using `Enumerable.Empty` It's not substantial, but it might be worth considering.

Reply:
DescendantTokens should never be null really. It will at least contain the name of the method it's invoking. So the enumerable.Empty should never be hit in all scenarios I can come up with. It's more as a safety guard in case a scenario we haven't thought of would crash the analyzer with a NRE without that null check.